### PR TITLE
Refactor modularization

### PR DESCRIPTION
**Pull request by j8r**
This opens the way for more various, independent module tests.
We can now test different parts of `Config`, without needing anything to set up.

### CODE REVIEW

File: src/actions.cr
Comment (dscottboggs) [inline src/actions.cr]: I like the idea here -- allowing overriding the block to change how access is granted (for testing I assume)... but I think the implementation could use some improvement. Instead of calling `#call` on the proc directly, I think the property should be called `access_filter` or something to that effect, has_access? can be like
```crystal
def has_access?(context, access)
  access_filter.call context, access
end
```
Secondly, I think the property should be on the authentication middleware, so we should move `DppmRestApi::Actions.auth_handler` to `DppmRestApi::Authentication.handler`. Then there can be an `Authentication.has_access?` next to it.

File: src/cli.cr
Comment (dscottboggs) [inline src/cli.cr:283]: Why?

File: src/config.cr
Comment (dscottboggs) [inline src/config.cr]: This would not get called when deserializing from JSON.

File: src/config/user.cr
Comment (dscottboggs) [inline src/config/user.cr]: What's the point of having a second copy of this?

File: src/dppm_rest_api.cr
Comment (dscottboggs) [inline src/dppm_rest_api.cr]: If we're going to make this change, we should have tests specifically for this proc, and in `spec_helper`, we should do `Actions.has_access = ->(context : HTTP::Server::Context, permission : Access) { true }` and expect a status code of 200.

**Review by dscottboggs — CHANGES_REQUESTED** [review]
I really like this change. Really good idea.

File: src/cli.cr
Comment (j8r) [reply src/cli.cr:283]: See https://crystal-lang.org/reference/syntax_and_semantics/return_types.html.

> Using `Void` is the same, but `Nil` is more idiomatic: `Void` is preferred in C bindings.

File: src/config/user.cr
Comment (j8r) [reply src/config/user.cr]: That's not a copy, array is a class, so passed as a reference.
How do you access to all groups from an `User` instance then (without depending to `DppmRestApi`)?

File: src/cli.cr
Comment (dscottboggs) [reply src/cli.cr:283]: Fair enough.

File: src/config/user.cr
Comment (dscottboggs) [reply src/config/user.cr]: Oh, ok, you're right. Since it's just a reference it's no big deal. I still don't know how to do this when deserializing from JSON though...

Comment (j8r) [reply src/config/user.cr]: I've changed the approach by using a `GroupView`

File: src/dppm_rest_api.cr
Comment (j8r) [reply src/dppm_rest_api.cr]: How to do this? I've tried something in `actions_spec.cr`

File: src/dppm_rest_api.cr
Comment (dscottboggs) [inline src/dppm_rest_api.cr:39]: I think I prefer to keep the middleware as a separate file at least. This section will only grow over time. Is there a reason we should prefer this form of keeping fewer files?

File: src/actions.cr
Comment (dscottboggs) [inline src/actions.cr:46]: :+1: :100:

**Review by dscottboggs — COMMENTED** [review]
This is really great. A couple things I would add:

```crystal
# In src/dppm_rest_api.cr
def default_acces_filter(context : HTTP::Server::Context, permission : Access) : Nil
  # (what you had)
end
def self.run(*args )
  run *args, &->default_access_filter(HTTP::Server::Context, Access)
end
def self.run(host : String, port : Int32, data_dir : String, &block : HTTP::Server::Context, Access -> Nil)
  # (what was already there, except)
  Actions.access_filter = ->block
end
```
Then, in `spec_helper`, we could do `DppmRestApi.run(host, port, __DIR__) { true }` to allow all requests for testing purposes.

File: src/dppm_rest_api.cr
Comment (j8r) [reply src/dppm_rest_api.cr:39]: Better having all the code at the same place instead of multiple arbitrary small files. This file is about 50 lines, which is totally fine.

We can move this away in the future if there are good reasons, for now I don't see any.

**Comment by j8r** [comment]
Done. Allow all requests in tests?

File: spec/spec_helper.cr
Comment (dscottboggs) [inline spec/spec_helper.cr:31]: Is this ok? If it is, I'm ready to merge this when you are.

**Review by dscottboggs — APPROVED** [review]

File: spec/spec_helper.cr
Comment (j8r) [reply spec/spec_helper.cr:31]: Sounds good. Any idea to test the filter in `actions_spec.cr`?
We can fix it in a further PR if we don't find a solutiom yet

Comment (dscottboggs) [reply spec/spec_helper.cr:31]: Well, the access_filter has to take several steps, each of which are already tested more or less. I agree that an integration/regression test on this would be due, but I think it would be better to set that up in a PR dedicated to integration and regression tests once the API is a little more fleshed out.

If you'd prefer, I could set up a test for this method before merging it.

Comment (j8r) [reply spec/spec_helper.cr:31]: I agree. I think we are good to merge, this will be simpler.

**Pull request merged**
